src/components/advanced-color-picker/advanced-color-picker.component.tsx

```diff
@@ -11,66 +10,122 @@ import { SimpleColorPicker, SimpleColor } from "../simple-color-picker";
+  value: string;
+}
+
+export interface AdvancedColorPickerPropTypes extends MarginProps {
+  /** Prop to specify the aria-describedby property of the component */
+  "aria-describedby"?: string;
+  /**
+   * Prop to specify the aria-label of the component.
+   * To be used only when the title prop is not defined, and the component is not labelled by any internal element.
+   */
+  "aria-label"?: string;
+  /**
+   * Prop to specify the aria-labeledby property of the component
+   * To be used when the title prop is a custom React Node,
+   * or the component is labelled by an internal element other than the title.
+   */
+  "aria-labelledby"?: string;
+  /** Prop for `availableColors` containing array of objects of colors */
+  availableColors: AdvancedColor[];
+  /** Prop for `defaultColor` containing the default color for `uncontrolled` use */
+  defaultColor: string;
+  /** Specifies the name prop to be applied to each color in the group */
+  name: string;
+  /** Prop for `onBlur` event */
+  onBlur?: (ev: React.FocusEvent<HTMLInputElement>) => void;
+  /** Prop for `onChange` event */
+  onChange?: (ev: React.ChangeEvent<HTMLInputElement>) => void;
+  /** Prop for `onClose` event */
+  onClose?: (ev: React.MouseEvent<HTMLElement>) => void;
+  /** Prop for `onOpen` event */
+  onOpen?: (ev: React.MouseEvent<HTMLElement>) => void;
+  /** Prop for `open` status */
+  open?: boolean;
+  /** The ARIA role to be applied to the component container */
+  role?: string;
+  /** Prop for `selectedColor` containing pre-selected color for `controlled` use */
+  selectedColor?: string;
+}
 
 const AdvancedColorPicker = ({
   "aria-describedby": ariaDescribedBy,
   "aria-label": ariaLabel,
   "aria-labelledby": ariaLabelledBy,
+  availableColors,
+  defaultColor,
   name,
-  open,
   onOpen,
   onClose,
   onChange,
   onBlur,
-  availableColors,
-  defaultColor,
-  selectedColor,
+  open,
   role,
+  selectedColor,
   ...props
-}) => {
+}: AdvancedColorPickerPropTypes) => {
   const isOpen = open || false;
-  const [dialogOpen, setDialogOpen] = useState();
+  const [colors, setColors] = useState<
+    { value: string; label: string; getRef: () => HTMLInputElement | null }[]
+  >();
+  const [dialogOpen, setDialogOpen] = useState<boolean>();
   const currentColor = selectedColor || defaultColor;
-  const [selectedColorRef, setSelectedColorRef] = useState();
-
-  const gridItemRefs = useRef(
-    Array.from(
-      {
-        length: availableColors.length,
-      },
-      () => React.createRef()
-    )
+  const [
+    selectedColorRef,
+    setSelectedColorRef,
+  ] = useState<HTMLInputElement | null>(null);
+
+  const simpleColorPickerData = useRef<{
+    gridItemRefs: Array<HTMLInputElement | null>;
+  }>();
+
+  useEffect(
+    () =>
+      setColors(
+        availableColors.map(({ value, label }, index) => {
+          return {
+            value,
+            label,
+            getRef: () =>
+              /* istanbul ignore next */
```

Comment:
this ternary is there only to satisfy the TypeScript compiler, it actually would never be null. If I change that to be your way, we will still have to ignore coverage.

Reply:
I'm ignoring the whole ternary, as it is there only to satisfy the TypeScript compiler as I said previously. If not for TypeScript, there would be nothing to cover in the first place as it would be just `simpleColorPickerData.current.gridItemRefs[index]`.

Reply:
I agree - we should ignore the `null` case. But you're ignoring both cases, which we don't want to do.

I'll explain it another way:

If we could write it as you want to, you would not put

```
() => /* istanbul ignore next */ simpleColorPickerData.current.gridItemRefs[index]
```

but that's effectively what you've got now.

To just ignore the `null` case, it should be:

```
() => simpleColorPickerData.current
  ? simpleColorPickerData.current.gridItemRefs[index]
  : /* istanbul ignore next */ null,
```

I hope you can see the difference now?

Reply:
yes I can, but I still see no reason to do it as we do not have any logic here that should have coverage. It looks more confusing to me like that.

Reply:
I'm just working on the principle of wanting to ignore as little code as possible. If we later accidentally remove all the tests that cover this`getRef` function (unlikely I know, but you should never say it's impossible), then with your placement of the ignore we would still keep 100% coverage.

Reply:
I think that the whole rule of having 100% coverage is not a good idea, but we don't have the other way to enforce people to cover their code, so we have to stick with that. In my opinion we should cover only the relevant code, not just every line just for the sake of having that 100%.

I have added a code comment to further explain to Carbon users, why this is ignored.